auth/oauth2.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/oauth2"
+)
+
+type Oauth2Authenticator struct {
+	TokenURL     string
+	ClientID     string
+	ClientSecret string
+	Username     string
+	Password     string
+
+	Token *oauth2.Token
+}
+
+func (o *Oauth2Authenticator) Configure(cfg map[string]interface{}) error {
+	decoded := struct {
+		TokenURL     string                 `mapstructure:"token_url"`
+		ClientID     string                 `mapstructure:"client_id"`
+		ClientSecret string                 `mapstructure:"client_secret"`
+		Username     string                 `mapstructure:"username"`
+		Password     string                 `mapstructure:"password"`
+		Rest         map[string]interface{} `mapstructure:",remain"`
+	}{}
+
+	if err := mapstructure.Decode(cfg, &decoded); err != nil {
+		return err
+	}
+
+	o.ClientID = decoded.ClientID
+	o.ClientSecret = decoded.ClientSecret
+	o.TokenURL = decoded.TokenURL
```

Comment:
Should we not check whether the URL is a valid URL before setting it?

Reply:
maybe `mapstructure:"token_url", valid:"url"` ?

Reply:
I'm not familiar with the `valid` tag, and by google-foo seems to be failing me. Could you provide a link to an explanation of how that works?

Reply:
Validation at the moment happends inside `validate()` which is called just below. While it may techincally be more go-ish ("gogonic"?) to not  until valudation, it would make the code less neat (we'd ether have to mix validation code with rest making the method less readable, or expose the currently interannly-defied temporary struct so that we can attach the validate method to it), and I don't see a concrete benefit?